pyriemann/classification.py

```diff
@@ -498,19 +500,113 @@ def fit(self, X, y):
         return self
 
     def predict(self, covtest):
-        """get the predictions.
+        """Get the predictions.
 
         Parameters
         ----------
-        X : ndarray, shape (n_trials, n_channels, n_channels)
-            ndarray of SPD matrices.
+        X : ndarray, shape (n_matrices, n_channels, n_channels)
+            Set of SPD matrices.
 
         Returns
         -------
-        pred : ndarray of int, shape (n_trials, 1)
+        pred : ndarray of int, shape (n_matrices, 1)
             the prediction for each trials according to the closest centroid.
         """
         dist = self._predict_distances(covtest)
         neighbors_classes = self.classes_[np.argsort(dist)]
         out, _ = stats.mode(neighbors_classes[:, 0:self.n_neighbors], axis=1)
         return out.ravel()
+
+
+class SupportVectorMachine(BaseEstimator, ClassifierMixin):
+    """Classification by Riemannian Support Vector Machine.
+
+    Support vector machine with precomputed Riemannian kernel matrix
+    according to different metrics as described in [1]_.
+
+    Parameters
+    ----------
+    svc_func : class, (default: sklearn.svm.SVC)
+        Class that performs the svc calculations.
+    metric : {'riemann', 'euclid', 'logeuclid'}
+        Metric for kernel matrix computation.
+    Cref : None | ndarray, shape (n_channels, n_channels)
+        Reference point for kernel matrix computation. If None, the mean of
+        the training data according to the metric is used.
+    **kwargs
+        Keyword arguments passed to svc_func.
+
+    Attributes
+    ----------
+    svc_ : svc_func instance
+        Fitted SVC with precomputed Riemannian kernel matrix.
+    data_ : ndarray, shape (n_matrices, n_channels, n_channels)
+        Training data.
+
+    Notes
+    -----
+    .. versionadded:: 0.2.8
+
+    References
+    ----------
+    .. [1] A. Barachant, S. Bonnet, M. Congedo, and C. Jutten.
+        Classification of covariance matrices using a Riemannian-based kernel
+        for BCI applications". In: Neurocomputing 112 (July 2013), pp. 172-178.
+    """
+
+    def __init__(self, svc_func=SVC, metric='riemann', Cref=None,
+                 **kwargs):
+        """Init."""
+        self.svc_func = svc_func
+        self.Cref = Cref
+        self.metric = metric
+        self.svc_params = kwargs
+        self.svc_ = self.svc_func(kernel='precomputed', **self.svc_params)
+
+    def __setattr__(self, name, value):
+        """Enable setting attributes for SVC subclass."""
+        if 'svc_' in self.__dict__.keys():
+            if name in self.svc_.__dict__.keys():
+                setattr(self.svc_, name, value)
+                return
+        super().__setattr__(name, value)
```

Comment:
you should rely on the set_params method from sklearn estimators. THis is a bit error prone.

Reply:
oh, didn't know about that. Will change.

Reply:
at least for the second part this doesn't work, because it also uses this method to set the parameters in the init.

Reply:
Is this still a problem if you separate SVC and SVR class?

Reply:
Yes, that doesn't change anything unfortunately.

Reply:
to me hacking __setattr__ is to me a code smell. We should try to avoid it. Maybe you can inherit from SVC and rely on `super()`'s constructor?

Reply:
Ah yes, I had tried it before and it didn't work. But now it works (had to copy all parameters from the SVC/SVR).